File: src/dir.c
#include <string.h>
#include <stdint.h>
#include "dir.h"
#include "kernal.h"

volatile uint8_t* border = (uint8_t*)(0xd020);

bool getbyte(uint8_t *chr)
{
    ++(*border);
    if (k_readst() != 0) return false;

    *chr = k_chrin();
    return true;
}

void opendir(fd_t fd, uint8_t dev)
{
    k_setnam("$");
    k_setlfs(fd, dev, 0);
    k_open();
    k_chkin(fd);

    uint8_t chr;
    // Skip target address
    getbyte(&chr);
    getbyte(&chr);
}

bool readdir(dirent *dirent)
{
    uint8_t chr;

    // Skip 2 bytes
    if (!getbyte(&chr)) return false;
    if (!getbyte(&chr)) return false;

    // BASIC line number
    if (!getbyte(&chr)) return false;
    if (!getbyte(&chr)) return false;

    // skip to opening quote
    while (chr != '"')
        if (!getbyte(&chr)) return false;

    // scan to closing quote
    char* name = dirent->d_name;

    if (!getbyte(&chr)) return false;
    while (chr != '"')
    {
        *(name++) = chr;
        if (!getbyte(&chr)) return false;
    }

    *name = '\0';

    // skip everything after the closing quote
    while (chr)
        if (!getbyte(&chr)) return false;

    return true;
}

void closedir(fd_t fd)
{
    k_close(fd);
    k_clrchn();
}

File: rs/src/dir.rs
use kernal::*;

fn get_byte () -> Option<u8> {
    unsafe {
        // TODO: flash border

        if k_readst() != 0 {
            None
        } else {
            Some(k_chrin())
        }
    }
}

pub fn open_dir (fd: u8, dev: u8) {
    unsafe {
        k_setnam("$".as_ptr());
        k_setlfs(fd, dev, 0);
        k_open();
        k_chkin(fd);

        // Skip target address
        get_byte();
        get_byte();
    }

    // Skip first directory entry, since that's just the disk label
    read_dir();
}

pub fn close_dir (fd: u8) {
    unsafe {
        k_close(fd);
        k_clrchn();
    }
}

#[derive(Clone,Copy)]
pub struct Dirent {
    pub d_name: [u8; 16 + 1]
}

pub fn read_dir () -> Option<Dirent> {
    // Skip 2 bytes
    get_byte()?;
    get_byte()?;

    // BASIC line number
    get_byte()?;
    get_byte()?;

    // Skip to opening quote
    while get_byte()? != 0x22 {};

    let mut i : usize = 0;
    let mut s : [u8; 16 + 1] = [0; 16+1];

    loop {
        let c = get_byte()?;
        if c == 0x22 { break };
        s[i] = c;
        i += 1;
    }

    while get_byte()? != 0x00 {};

    Some(Dirent{
        d_name: s
    })
}
